File: cpp/526.beautiful-arrangement.cpp
/*
 * @lc app=leetcode id=526 lang=cpp
 *
 * [526] Beautiful Arrangement
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    int countArrangement(int n)
    {
        int res = 0;
        vector<bool> used(n);
        backtrack(res, n, used);
        return res;
    }
    void backtrack(int &res, int n, vector<bool> &used)
    {
        if (n == 0)
        {
            ++res;
            return;
        }
        int index = 1;
        for (auto val : used)
        {
            if (!val && (!(n % index) || !(index % n)))
            {
                val = true;
                backtrack(res, n - 1, used);
                val = false;
            }
            ++index;
        }
    }
};
// @lc code=end

File: java/526.beautiful-arrangement.java
/*
 * @lc app=leetcode id=526 lang=java
 *
 * [526] Beautiful Arrangement
 */

// @lc code=start
class Solution {
    int res = 0;

    public int countArrangement(int n) {
        boolean[] seen = new boolean[n];
        backtrack(n, seen);
        return res;
    }

    void backtrack(int n, boolean[] seen) {
        if (n == 0) {
            ++res;
            return;
        }
        for (int i = 0; i < seen.length; ++i) {
            if (!seen[i] && ((n % (i + 1)) == 0 || ((i + 1) % n) == 0)) {
                seen[i] = true;
                backtrack(n - 1, seen);
                seen[i] = false;
            }
        }
    }
}
// @lc code=end
